### DISCUSSION TITLE
Twilio->Deepgram live streaming in Node.js keeps disconnecting over and over

### DISCUSSION THREAD

**Discussion by c0bra** [deepgram/community#565] [General help] [answered]
I've attempted to convert the Python code from the Twilio->Deepgram live streaming transcription into Node.js  After several hundred attempts I keep getting stuck with the websocket connection to Deepgram getting closed over and over.

I will occasionally get a couple words transcribed, but mostly I am able to send a chunk or two, get an empty transccription back, and then it closes with a message saying `[1000] Normal connection closure`:

```bash
> node index.js
Websocket server listening on port 6000

Connecting to Deepgram...
New connection from  ::ffff:127.0.0.1
Transcript: [ { transcript: '', confidence: 0, words: [] } ]
Transcript: [ { transcript: '', confidence: 0, words: [] } ]
Deepgram connection closed: [1000] Normal connection closure
socket: data couldn't be sent to deepgram
socket: retrying connection to deepgram
Connecting to Deepgram...
Transcript: [ { transcript: '', confidence: 0, words: [] } ]
Transcript: [ { transcript: '', confidence: 0, words: [] } ]
Deepgram connection closed: [1000] Normal connection closure
socket: data couldn't be sent to deepgram
socket: retrying connection to deepgram
Connecting to Deepgram...
```

I am not sending null bytes AFAICT; the bytelength on the Buffer I am sending is good, so I am not expecting that I would be implicitly closing the websocket on accident.

I am out of ideas here and am about to try to do this in another language if I can't get some insights from this community. Appreciate your help!

**Edit:** [Gist here](https://gist.github.com/c0bra/bacadd4ed1de2d18d58307458f403d66)

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Reply by nikolawhallon** [reply, marked as answer]
There is a node js example here that was recently updated to a newer version of the SDK: https://github.com/deepgram/public-code-examples/tree/master/twilio/node

it doesn't do any corrections based on missing packets from Twilio like the Python example does (and like you are trying to convert), but it this works out of the box, maybe that's enough to get started (and honestly you can ignore dropped packets from Twilio unless you start to notice major timestamp deviations).

**Reply by c0bra** [reply]
@nikolawhallon Thanks for the tip! That samples seems to make sense. 

**Edit**: (Removed a question that I should have spent longer trying to answer)

I was able to convert the python code to Go and get it working but I may try this one too.

**Reply by nikolawhallon** [reply]
The code there is interleaving the two channels and this gets sent to Deepgram as a single stream (with two channels interleaved sample by sample). This is the format for multichannel raw audio.

**Reply by c0bra** [reply]
Opened some PRs based on my testing of the stereo example.

**Reply by ty-chris** [reply]
Hi there,

Sorry to revive this closed topic, I am currently facing the same issue where I potentially send a few audio buffers simultaneously to the Deepgram Websocket Endpoint. 

When just using a single audio stream from a client to my server which handles the forwarding of audio to deepgram websocket endpoint, everything works as expected. However, the moment a second audio stream is included, the error message: "deepgram: disconnected Normal connection closure" is received. Is there more information as to what could cause deepgram to do a "Normal connection closure"?

The way my app/server is setup is as follows:

Each client has a websocket connection to my server.

My server receives each audio stream from each client. As the audio stream is received, it then sends it all to a single deepgram websockets endpoint. The expected result is that based on the time each audio stream is sent out, that we could have a collaborative transcript in real-time. However, this does not work as expected.

Wondering if I could get more information regarding what could cause the Normal connection closure on deepgram's end so I could debug this. Thank you!

**Reply by JasonLeviGoodison** [reply]
+1 I have a similar error. I placed Deepgram behind an express server, and then I have a client which is streaming multiple files. Same thing happens on my end. I think its because each "send" is not sequential - but have no way of really knowing. There is no API Doc on a "normal connection closure".

**Reply by ldenoue** [reply]
@ty-chris and @JasonLeviGoodison I don't think that Deepgram is designed to handle several audio sent to the websocket. how would you even be able to know which words correspond to which audio? Or are you saying there's no overlap between these different audio streams? In any case, I think you could use a queue to add your audio data, and have the deepgram websocket dequeue data sequentially from that queue.
